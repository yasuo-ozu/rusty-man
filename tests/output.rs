// SPDX-FileCopyrightText: 2020 Robin Krahl <robin.krahl@ireas.org>
// SPDX-License-Identifier: MIT

#[path = "../src/test_utils.rs"]
mod test_utils;

use std::env;
use std::path;
use std::process;

use assert_cmd::prelude::*;

use test_utils::{with_rustdoc, Format};

fn run(path: impl AsRef<path::Path>, args: &[&str]) -> assert_cmd::assert::Assert {
    process::Command::cargo_bin(env!("CARGO_PKG_NAME"))
        .unwrap()
        .args(&["--no-default-sources", "--source"])
        .arg(path.as_ref())
        .args(&["--viewer", "plain"])
        .args(&["--width", "100"])
        .args(args)
        .assert()
}

fn get_stdout(path: impl AsRef<path::Path>, args: &[&str]) -> String {
    let cmd = run(path, args).success().stderr("");
    String::from_utf8(cmd.get_output().stdout.clone()).unwrap()
}

macro_rules! generate_run {
    ($name:ident $version:literal $formats:expr; $args:expr) => {
        #[test]
        fn $name() {
            with_rustdoc($version, $formats, |version, format, path| {
                insta::assert_snapshot!(
                    format!("{}_{}_{}", version, format, std::stringify!($name)),
                    get_stdout(path, $args)
                );
            });
        }
    };
}

macro_rules! assert_doc {
    ($( $name:ident($version:literal, $formats:expr): $item:literal ),* $(,)? ) => {
        $(
            generate_run!($name $version $formats; &[$item]);
        )*
    };
}

macro_rules! assert_examples {
    ($( $name:ident($version:literal, $formats:expr): $item:literal ),* $(,)? ) => {
        $(
            generate_run!($name $version $formats; &["-e", $item]);
        )*
    };
}

assert_doc![
    mod_anyhow("*", Format::all()): "anyhow",
    macro_anyhow_anyhow("*", Format::all()): "anyhow::anyhow",
    macro_anyhow_ensure("*", Format::all()): "anyhow::ensure",
    struct_anyhow_error("<1.51.0", Format::all()): "anyhow::Error",
    trait_anyhow_context("*", Format::all()): "anyhow::Context",
    type_anyhow_result("*", Format::all()): "anyhow::Result",
];

assert_doc![
    mod_log("*", Format::all()): "log",
    macro_log_debug("*", Format::all()): "log::debug",
    struct_log_metadata("*", Format::all()): "log::Metadata",
    enum_log_level("*", Format::all()): "log::Level",
    constant_log_static_max_level("*", Format::all()): "log::STATIC_MAX_LEVEL",
    trait_log_log("*", Format::all()): "log::Log",
    fn_log_logger("*", Format::all()): "log::logger",
    fn_log_set_logger_racy("*", Format::all()): "log::set_logger_racy",
];

assert_doc![
    mod_rand_core("*", Format::all()): "rand_core",
    trait_rand_core_rngcore("*", Format::all()): "rand_core::RngCore",
    trait_rand_core_seedablerng("*", Format::all()): "rand_core::SeedableRng",
    struct_rand_core_block_blockrng("*", Format::all()): "rand_core::block::BlockRng",
];

assert_examples![
    examples_mod_anyhow("*", Format::all()): "anyhow",
    examples_mod_log(">1.40.0", Format::all()): "log",
    examples_struct_rand_core_rngcore("*", Format::all()): "rand_core::RngCore",
];
