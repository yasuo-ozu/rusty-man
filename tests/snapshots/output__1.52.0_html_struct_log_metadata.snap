---
source: tests/output.rs
expression: "get_stdout(path, &[\"log::Metadata\"])"
---
log                                     Struct log::Metadata                               rusty-man

SYNOPSIS
      pub struct Metadata<'a> { /* fields omitted */ }

DESCRIPTION
      Metadata about a log message.
      
      # Use
      
      `Metadata` structs are created when users of the library use logging macros.
      
      They are consumed by implementations of the `Log` trait in the `enabled` method.
      
      `Record`s use `Metadata` to determine the log message’s severity and target.
      
      Users should use the `log_enabled!` macro in their code to avoid constructing expensive log
      messages.
      
      # Examples
      
      ⓘ
      use log::{Record, Level, Metadata};
      struct MyLogger;
      impl log::Log for MyLogger {
          fn enabled(&self, metadata: &Metadata) -> bool {
              metadata.level() <= Level::Info
          }
          fn log(&self, record: &Record) {
              if self.enabled(record.metadata()) {
                  println!("{} - {}", record.level(), record.args());
              }
          }
          fn flush(&self) {}
      }

METHODS
   impl<'a> Metadata<'a>
      builder
            pub fn builder() -> MetadataBuilder<'a>

            Returns a new builder.

      level
            pub fn level(&self) -> Level

            The verbosity level of the message.

      target
            pub fn target(&self) -> &'a str

            The name of the target of the directive.

IMPLEMENTATIONS
   Trait Implementations
      Clone
            impl<'a> Clone for Metadata<'a>

      Debug
            impl<'a> Debug for Metadata<'a>

      Eq
            impl<'a> Eq for Metadata<'a>

      Hash
            impl<'a> Hash for Metadata<'a>

      Ord
            impl<'a> Ord for Metadata<'a>

      PartialEq
            impl<'a> PartialEq<Metadata<'a>> for Metadata<'a>

      PartialOrd
            impl<'a> PartialOrd<Metadata<'a>> for Metadata<'a>

      StructuralEq
            impl<'a> StructuralEq for Metadata<'a>

      StructuralPartialEq
            impl<'a> StructuralPartialEq for Metadata<'a>

   Auto Trait Implementations
      RefUnwindSafe
            impl<'a> RefUnwindSafe for Metadata<'a>

      Send
            impl<'a> Send for Metadata<'a>

      Sync
            impl<'a> Sync for Metadata<'a>

      Unpin
            impl<'a> Unpin for Metadata<'a>

      UnwindSafe
            impl<'a> UnwindSafe for Metadata<'a>

   Blanket Implementations
      Any
            impl<T> Any for T 
            where
                T: 'static + ?Sized,

      Borrow
            impl<T> Borrow<T> for T 
            where
                T: ?Sized,

      BorrowMut
            impl<T> BorrowMut<T> for T 
            where
                T: ?Sized,

      From
            impl<T> From<T> for T

      Into
            impl<T, U> Into<U> for T 
            where
                U: From<T>,

      ToOwned
            impl<T> ToOwned for T 
            where
                T: Clone,

      TryFrom
            impl<T, U> TryFrom<U> for T 
            where
                U: Into<T>,

      TryInto
            impl<T, U> TryInto<U> for T 
            where
                U: TryFrom<T>,


